File: leetcode/solutions/range-sum-of-bst.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def rangeSumBST(self, root: TreeNode, low: int, high: int) -> int:
        """
        :type root: TreeNode
        :type low: int
        :type high: int
        :rtype: int
        """
        def dfs(node):
            nonlocal ans
            if node:
                if low <= node.val <= high:
                    ans += node.val
                if low < node.val:
                    dfs(node.left)
                if high > node.val:
                    dfs(node.right)
        ans = 0
        dfs(root)
        return ans

File: leetcode/solutions/range-sum-of-bst.swift
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public var val: Int
 *     public var left: TreeNode?
 *     public var right: TreeNode?
 *     public init() { self.val = 0; self.left = nil; self.right = nil; }
 *     public init(_ val: Int) { self.val = val; self.left = nil; self.right = nil; }
 *     public init(_ val: Int, _ left: TreeNode?, _ right: TreeNode?) {
 *         self.val = val
 *         self.left = left
 *         self.right = right
 *     }
 * }
 */
class Solution {
    func rangeSumBST(_ root: TreeNode?, _ low: Int, _ high: Int) -> Int {
        func rangeSumRecu(node: TreeNode, low: Int, high: Int, sum: inout Int) {
            if node.val >= low, node.val <= high {
                sum += node.val
            }
            if low < node.val, let node = node.left {
                rangeSumRecu(node: node, low: low, high: high, sum: &sum)
            }
            if high > node.val, let node = node.right {
                rangeSumRecu(node: node, low: low, high: high, sum: &sum)
            }

        }
        var sum = 0
        guard let root = root else {
            return sum
        }
        rangeSumRecu(node: root, low: low, high: high, sum: &sum)
        return sum
    }
}
